src/main/java/com/google/cloud/genomics/dataflow/utils/PairGenerator.java

```diff
@@ -27,10 +26,26 @@
 import com.google.common.collect.Iterators;
 import com.google.common.collect.Range;
 
+/**
+ * Generates all combinations (not permutations) of pairs of the elements in the iterable.
+ *
+ */
 public class PairGenerator {
 
+  /**
+   * Generates all combinations (not permutations) of pairs of the elements in the iterable.
+   * 
+   * The pairs are often used as keys for results in many n^2 analyses where we end up with
+   * (n*(n-1))/2 or ((n*(n-1))/2)+n results.
+   * 
+   * @param list
+   * @param withReplacement - When true an additional N pairs will be generated (one for each
+   *        element in the list) in addition to the Nchoose2 pairs.
+   * @param comparator - used to enforce an order upon the elements within each pair
+   * @return
+   */
   public static <X, L extends List<? extends X> & RandomAccess> FluentIterable<KV<X, X>> allPairs(
-      final L list) {
+      final L list, final boolean withReplacement, final Comparator<X> comparator) {
```

Comment:
https://sites.google.com/a/google.com/java-practices/boolean-parameters

I suggest making PairGenerator an enum with two values:

    public enum PairGenerator {
    
      WITH_REPLACEMENT(true),
      WITHOUT_REPLACEMENT(false);
    
      private final boolean withReplacement;
    
      private PairGenerator(boolean withReplacement) {
        this.withReplacement = withReplacement;
      }
    
      public <X, L extends List<? extends X> & RandomAccess> FluentIterable<KV<X, X>> allPairs(final L list, final Comparator<? super X> comparator) {
        return FluentIterable
            .from(ContiguousSet.create(Range.closedOpen(0, list.size()), DiscreteDomain.integers()))
            .transformAndConcat(
                new Function<Integer, Iterable<KV<X, X>>>() {
                  @Override public Iterable<KV<X, X>> apply(final Integer i) {
                    return new Iterable<KV<X, X>>() {
                          @Override public Iterator<KV<X, X>> iterator() {
                            int iteratorIndex = i;
                            if (!withReplacement) {
                              iteratorIndex++;
                            }
                            return Iterators.transform(
                                list.listIterator(iteratorIndex),
                                new Function<X, KV<X, X>>() {
    
                                  private final X key = list.get(i);
    
                                  @Override public KV<X, X> apply(X value) {
                                    boolean swap = 0 < comparator.compare(key, value);
                                    return KV.of(swap ? value : key, swap ? key : value);
                                  }
                                });
                          }
                        };
                  }
                });
      }
    }

Notice that we can generalize the type of comparator by making it contravariant  in its parameter X
Also, we can save a few autounboxings by declaring the type of interatorIndex to be int instead of Integer